presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LeftJoinNullFilterToSemiJoin.java

```diff
@@ -0,0 +1,150 @@
+    {
+        return expression instanceof SpecialFormExpression && ((SpecialFormExpression) expression).getForm().equals(SpecialFormExpression.Form.IS_NULL)
+                && ((SpecialFormExpression) expression).getArguments().get(0).equals(rightKey);
+    }
+
+    @Override
+    public Result apply(FilterNode filterNode, Captures captures, Context context)
+    {
+        JoinNode joinNode = captures.get(CHILD);
+        VariableReferenceExpression rightKey = joinNode.getCriteria().get(0).getRight();
+        VariableReferenceExpression leftKey = joinNode.getCriteria().get(0).getLeft();
+        // if output other than join key from right side are used, cannot rewrite
+        if (joinNode.getOutputVariables().stream().anyMatch(x -> joinNode.getRight().getOutputVariables().contains(x) && !x.equals(rightKey))) {
+            return Result.empty();
+        }
+
+        List<RowExpression> andConjuncts = extractConjuncts(filterNode.getPredicate());
+        List<RowExpression> rightKeyNull = andConjuncts.stream().filter(x -> isRightKeyNullExpression(x, rightKey)).collect(toImmutableList());
+        // if no null check just return
+        if (rightKeyNull.isEmpty()) {
+            return Result.empty();
+        }
+        List<RowExpression> remainingConjunct = andConjuncts.stream().filter(x -> !rightKeyNull.contains(x)).collect(toImmutableList());
+        List<VariableReferenceExpression> variablesInRemainingConjuncts = remainingConjunct.stream().flatMap(x -> extractAll(x).stream()).collect(toImmutableList());
+        // Variables from right side referred in other expressions, not only null check.
+        if (variablesInRemainingConjuncts.stream().anyMatch(x -> joinNode.getBuild().getOutputVariables().contains(x))) {
+            return Result.empty();
+        }
+
+        VariableReferenceExpression semiJoinOutput = context.getVariableAllocator().newVariable("semiJoinOutput", BOOLEAN);
+        SemiJoinNode semiJoinNode = new SemiJoinNode(filterNode.getSourceLocation(), context.getIdAllocator().getNextId(), joinNode.getProbe(), joinNode.getBuild(),
+                leftKey, rightKey, semiJoinOutput, Optional.empty(), Optional.empty(), Optional.empty(), ImmutableMap.of());
+        RowExpression filterExpression = not(functionAndTypeManager, coalesceNullToFalse(semiJoinOutput));
+        FilterNode semiFilterNode = new FilterNode(filterNode.getSourceLocation(), context.getIdAllocator().getNextId(), semiJoinNode, filterExpression);
+        Map<VariableReferenceExpression, RowExpression> outputAssignments =
+                filterNode.getOutputVariables().stream().collect(toImmutableMap(identity(), x -> x.equals(rightKey) ? constantNull(rightKey.getType()) : x));
+        PlanNode planNode = new ProjectNode(context.getIdAllocator().getNextId(), semiFilterNode, Assignments.builder().putAll(outputAssignments).build());
+        if (!remainingConjunct.isEmpty()) {
```

Comment:
You need to make sure the remaining conjuncts are all on the lhs.

Reply:
It's checked above
```
        List<VariableReferenceExpression> variablesInRemainingConjuncts = remainingConjunct.stream().flatMap(x -> extractAll(x).stream()).collect(toImmutableList());
        // Variables from right side referred in other expressions, not only null check.
        if (variablesInRemainingConjuncts.stream().anyMatch(x -> joinNode.getRight().getOutputVariables().contains(x))) {
            return Result.empty();
        }
```